File: data_analysis/scripts/features.py
"""P3 feature definitions matching firmware/src/features.cpp."""

from dataclasses import asdict, dataclass, replace
import math

import numpy as np

from fft_processor import spectral_features

FEATURE_NAMES = ("rms", "peak_to_peak", "crest_factor",
                 "dominant_frequency", "band_energy")


@dataclass(frozen=True)
class FeatureConfig:
    sample_rate_hz: float = 800.0
    window_size: int = 512
    band_low_hz: float = 10.0
    band_high_hz: float = 200.0
    remove_mean: bool = True

    def validate(self):
        count = self.window_size
        if not isinstance(count, int) or count < 4 or count > 1024 or count & (count - 1):
            raise ValueError("window_size phải là lũy thừa của 2, từ 4 đến 1024.")
        if not math.isfinite(self.sample_rate_hz) or self.sample_rate_hz <= 0:
            raise ValueError("sample_rate_hz phải hữu hạn và lớn hơn 0.")
        if not (math.isfinite(self.band_low_hz) and math.isfinite(self.band_high_hz)
                and 0 <= self.band_low_hz < self.band_high_hz <= self.sample_rate_hz / 2):
            raise ValueError("Dải tần phải thỏa 0 <= low < high <= Fs/2.")

    def with_sample_rate(self, rate):
        return replace(self, sample_rate_hz=rate)

    def to_dict(self):
        return asdict(self)


def extract_features(samples, config):
    config.validate()
    values = np.asarray(samples, dtype=float)
    if values.ndim != 1 or len(values) != config.window_size or not np.all(np.isfinite(values)):
        raise ValueError("Cửa sổ phải đủ số mẫu hữu hạn.")
    if config.remove_mean:
        values = values - np.mean(values)
    rms = float(np.sqrt(np.mean(values ** 2)))
    dominant, energy = spectral_features(values, config.sample_rate_hz,
                                         config.band_low_hz, config.band_high_hz)
    result = {
        "rms": rms,
        "peak_to_peak": float(np.ptp(values)),
        "crest_factor": float(np.max(np.abs(values)) / rms) if rms > 0 else 0.0,
        "dominant_frequency": dominant,
        "band_energy": energy,
    }
    if not all(math.isfinite(value) and value >= 0 for value in result.values()):
        raise ValueError("Đặc trưng không hợp lệ.")
    return result

File: firmware/src/features.cpp
// Owner: P3
#include "features.h"

#include <algorithm>
#include <cmath>

#include "fft_processor.h"

namespace {

FeatureConfig activeConfig{10.0f, 200.0f, true};

bool validWindow(const float* samples, std::size_t count, float sampleRateHz) {
    return samples != nullptr && count >= 4 && count <= P3_MAX_SAMPLE_COUNT &&
           (count & (count - 1U)) == 0U && std::isfinite(sampleRateHz) &&
           sampleRateHz > 0.0f && activeConfig.bandHighHz <= sampleRateHz / 2.0f;
}

FanVibrationFeatures invalidFeatures() {
    return {NAN, NAN, NAN, NAN, NAN};
}

}  // namespace

bool configureFeatures(const FeatureConfig& config) {
    if (!std::isfinite(config.bandLowHz) ||
        !std::isfinite(config.bandHighHz) || config.bandLowHz < 0.0f ||
        config.bandHighHz <= config.bandLowHz) {
        return false;
    }
    activeConfig = config;
    return true;
}

bool areFeaturesValid(const FanVibrationFeatures& features) {
    const float values[] = {
        features.rms, features.peakToPeak, features.crestFactor,
        features.dominantFrequency, features.bandEnergy};
    for (float value : values) {
        if (!std::isfinite(value) || value < 0.0f) {
            return false;
        }
    }
    return true;
}

bool extractFanFeatures(const float* samples, std::size_t count,
                        float sampleRateHz, FanVibrationFeatures& output) {
    output = invalidFeatures();
    if (!validWindow(samples, count, sampleRateHz)) {
        return false;
    }

    double total = 0.0;
    for (std::size_t index = 0; index < count; ++index) {
        if (!std::isfinite(samples[index])) {
            return false;
        }
        total += samples[index];
    }
    const float mean = activeConfig.removeMean
                           ? static_cast<float>(total / count)
                           : 0.0f;

    double sumSquares = 0.0;
    float minimum = INFINITY;
    float maximum = -INFINITY;
    float peak = 0.0f;
    for (std::size_t index = 0; index < count; ++index) {
        const float value = samples[index] - mean;
        sumSquares += static_cast<double>(value) * value;
        minimum = std::min(minimum, value);
        maximum = std::max(maximum, value);
        peak = std::max(peak, std::fabs(value));
    }

    const float rms = static_cast<float>(std::sqrt(sumSquares / count));
    SpectralFeatures spectral{};
    if (!processFFT(samples, count, sampleRateHz, mean,
                    activeConfig.bandLowHz, activeConfig.bandHighHz, spectral)) {
        return false;
    }

    output = {rms, maximum - minimum, rms > 0.0f ? peak / rms : 0.0f,
              spectral.dominantFrequency, spectral.bandEnergy};
    return areFeaturesValid(output);
}

bool extractFeatures(const float* samples, std::size_t count,
                     float sampleRateHz, VibrationFeatures& output) {
    FanVibrationFeatures full{};
    if (!extractFanFeatures(samples, count, sampleRateHz, full)) {
        output = {NAN, NAN, NAN};
        return false;
    }
    output = {full.rms, full.peakToPeak, full.crestFactor};
    return true;
}
